v2/pkg/engine/datasource/graphql_datasource/graphql_tws_handler.go

```diff
@@ -1,88 +1,166 @@
+	err := wsutil.WriteClientText(h.conn, []byte(req))
+	if err != nil {
+		h.log.Error("failed to write complete message", log.Error(err))
+	}
+	_ = h.conn.Close()
+}
+
+func (h *gqlTWSConnectionHandler) Subscribe() error {
+	return h.subscribe()
+}
+
+func (h *gqlTWSConnectionHandler) ReadMessage() (done, timeout bool) {
+
+	r := bufio.NewReader(h.conn)
+	wr := bufio.NewWriter(h.conn)
+	rwr := bufio.NewReadWriter(r, wr)
+
+	for {
+
+		err := h.conn.SetReadDeadline(time.Now().Add(time.Second))
+		if err != nil {
+			return h.handleConnectionError(err)
+		}
+
+		data, err := wsutil.ReadServerText(rwr)
+		if err != nil {
+			return h.handleConnectionError(err)
+		}
+
+		messageType, err := jsonparser.GetString(data, "type")
+		if err != nil {
+			return false, false
+		}
+		switch messageType {
+		case messageTypePing:
+			h.handleMessageTypePing()
+			continue
+		case messageTypeNext:
+			h.handleMessageTypeNext(data)
+			continue
+		case messageTypeComplete:
+			h.handleMessageTypeComplete(data)
+			return true, false
+		case messageTypeError:
+			h.handleMessageTypeError(data)
+			continue
+		case messageTypeConnectionKeepAlive:
+			continue
+		case messageTypeData, messageTypeConnectionError:
+			h.log.Error("Invalid subprotocol. The subprotocol should be set to graphql-transport-ws, but currently it is set to graphql-ws")
+			return true, false
+		default:
+			h.log.Error("unknown message type", log.String("type", messageType))
+			return false, false
+		}
+	}
+}
+
+func (h *gqlTWSConnectionHandler) handleConnectionError(err error) (closed, timeout bool) {
+	if errors.Is(err, context.DeadlineExceeded) {
+		return false, true
+	}
+	netOpErr := &net.OpError{}
+	if errors.As(err, &netOpErr) {
+		if netOpErr.Timeout() {
+			return false, true
+		}
+		return true, false
+	}
+	if errors.As(err, &wsutil.ClosedError{}) {
+		return true, false
+	}
+	if strings.HasSuffix(err.Error(), "use of closed network connection") {
+		return true, false
+	}
+	return false, false
 }
 
-func newGQLTWSConnectionHandler(ctx context.Context, conn *websocket.Conn, rt time.Duration, l log.Logger) *gqlTWSConnectionHandler {
+func (h *gqlTWSConnectionHandler) NetConn() net.Conn {
+	return h.conn
+}
+
+func newGQLTWSConnectionHandler(requestContext, engineContext context.Context, conn net.Conn, options GraphQLSubscriptionOptions, updater resolve.SubscriptionUpdater, l log.Logger) *gqlTWSConnectionHandler {
 	return &gqlTWSConnectionHandler{
-		conn:               conn,
-		ctx:                ctx,
-		log:                l,
-		nextSubscriptionID: 0,
-		subscriptions:      map[string]Subscription{},
-		readTimeout:        rt,
+		conn:           conn,
+		requestContext: requestContext,
+		engineContext:  engineContext,
+		log:            l,
+		updater:        updater,
+		options:        options,
 	}
 }
 
-func (h *gqlTWSConnectionHandler) StartBlocking(sub Subscription) {
-	readCtx, cancel := context.WithCancel(h.ctx)
+func (h *gqlTWSConnectionHandler) StartBlocking() error {
```

Comment:
Why do we still need this logic? Is this for Windows support?

Reply:
correct, Windows doesn't use epoll, so it will use the blocking approach with a goroutine.